runtime/compiler/p/codegen/J9TreeEvaluator.cpp

```diff
@@ -11797,6 +11797,355 @@ static TR::Register *inlineEncodeUTF16(TR::Node *node, TR::CodeGenerator *cg)
+   generateConditionalBranchInstruction(cg, TR::InstOpCode::bne, node, foundLabel, cr6);
+
+   generateLabelInstruction(cg, TR::InstOpCode::label, node, notFoundLabel);
+
+   // We've looked through the entire string and didn't find our target character, so return the
+   // sentinel value -1
+   generateTrg1ImmInstruction(cg, TR::InstOpCode::li, node, result, -1);
+   generateLabelInstruction(cg, TR::InstOpCode::b, node, endLabel);
+
+   generateLabelInstruction(cg, TR::InstOpCode::label, node, foundLabel);
+
+   // We've managed to find a match for the target value in the loaded vector, but we don't yet know
+   // which element of the loaded vector is the first match. The comparison will have set matching
+   // elements in the vector to -1 and non-matching elements to 0. We can find the first matching
+   // element by gathering the first bit of every byte in the vector register...
+
+   // Set permuteVector = 0x000102030405060708090a0b0c0d0e0f
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::lvsl, node, permuteVector, zeroRegister, zeroRegister);
+
+   // For little-endian, reverse permuteVector so that we can find the first set bit using a count
+   // leading zeroes test instead of a count trailing zeroes test. This is necessary since cnttzw
+   // wasn't introduced until Power 9.
+   if (TR::Compiler->target.cpu.isLittleEndian())
+      {
+      generateTrg1ImmInstruction(cg, TR::InstOpCode::vspltisb, node, targetVector, 0x0f);
+      generateTrg1Src2Instruction(cg, TR::InstOpCode::vsububm, node, permuteVector, targetVector, permuteVector);
+      }
+
+   // Set permuteVector = 0x00081018202830384048505860687078 (reversed for LE)
+   generateTrg1ImmInstruction(cg, TR::InstOpCode::vspltisb, node, targetVector, 3);
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::vslb, node, permuteVector, permuteVector, targetVector);
+
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::vbpermq, node, searchVector, searchVector, permuteVector);
+   generateTrg1Src1Instruction(cg, TR::InstOpCode::mfvsrwz, node, result, searchVector);
+
+   // Then count the number of leading zeroes from the obtained result. This tells us the index (in
+   // bytes) of the first matching element in the vector. Note that there is no way to count leading
+   // zeroes of a half-word, so we count leading zeroes of a word and subtract 16 since the value
+   // we're interested in is in the least significant half-word.
+   generateTrg1Src1Instruction(cg, TR::InstOpCode::cntlzw, node, result, result);
+   generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::addi, node, result, result, -16);
+
+   // Finally, combine this with the address of the last vector load to find the address of the
+   // first matching element in the string. Finally, use this to calculate the corresponding index.
+   generateTrg1Src2Instruction(cg, TR::InstOpCode::add, node, result, result, currentAddress);
```

Comment:
When the match is found in the first vector, the byte-index should be (result - (arrAddress & 0xF)) disregarding currentAddress. (result + (currentAddress - arrAddress))  is only correct for matching in none first vector.  A simple fix is to add one instruction before this add:  currentAddress = currentAddress & ~(0xF);

Reply:
I think you've misunderstood the part of the code responsible for loading the first misaligned vector. This is the correct way to calculate the address of the match because the data in `searchVector` has _already_ been shifted backwards by `currentAddress & 0xf` bytes earlier by the code responsible for replacing the garbage loaded before the string data. Because of this, the first byte of `searchVector` corresponds to the byte at `(currentAddress & ~0xf) + (currentAddress & 0xf) = currentAddress` _not_ the first byte that was actually loaded from memory (which would be at `currentAddress & ~0xf`).

Reply:
Fair enough ...